In [2]:
import numpy as np
import pandas as pd

# Load du df depuis BQ

In [3]:
from google.cloud import bigquery
from google.oauth2 import service_account
import sys; sys.path.append('..')
from params import PROJECT_ID, TABLE_ID

def load_dataset(PROJECT_ID, TABLE_ID):

    print(PROJECT_ID, TABLE_ID)

    TABLE_ID = 'test_dataset.test10'

    credentials = service_account.Credentials.from_service_account_file("../bq_keys.json")
    #credentials = service_account.Credentials.from_service_account_file('~//Users//admin/code//tloizel//wagon-give-me-a-sign//bq_keys.json')

    client = bigquery.Client(project=PROJECT_ID, credentials=credentials)
    table = client.get_table(TABLE_ID)


    # Construct the SQL query to retrieve the table data
    query = f'SELECT * FROM `{table}`'

    # Submit the query and fetch the results
    df = client.query(query).to_dataframe()
    
    return df

In [4]:
df = load_dataset(PROJECT_ID, TABLE_ID)
df.head(3)

wagon-give-me-a-sign test_dataset.test26


x_0  y_0  z_0       x_1       y_1       z_1       x_2       y_2       z_2  \
0  0.0  0.0  0.0  0.068996 -0.001794 -0.026495  0.129794 -0.049328 -0.037758   
1  0.0  0.0  0.0  0.084189 -0.004527 -0.035077  0.159894 -0.063521 -0.049802   
2  0.0  0.0  0.0  0.081866  0.003673 -0.039060  0.166663 -0.060628 -0.056427   

        x_3  ...      x_18      y_18      z_18      x_19      y_19      z_19  \
0  0.157946  ...  0.015763 -0.236859 -0.053696  0.014897 -0.160986 -0.042879   
1  0.195749  ...  0.022846 -0.301951 -0.067628  0.019003 -0.205834 -0.053242   
2  0.204637  ...  0.011525 -0.315707 -0.082841  0.011519 -0.209255 -0.069801   

       x_20      y_20      z_20  target  
0  0.003309 -0.143430 -0.028293       a  
1  0.004331 -0.182658 -0.034498       a  
2 -0.000003 -0.184926 -0.051607       a  

[3 rows x 64 columns]

In [5]:
df[df['target'] == 'a']

x_0  y_0  z_0       x_1       y_1       z_1       x_2       y_2  \
0    0.0  0.0  0.0  0.068996 -0.001794 -0.026495  0.129794 -0.049328   
1    0.0  0.0  0.0  0.084189 -0.004527 -0.035077  0.159894 -0.063521   
2    0.0  0.0  0.0  0.081866  0.003673 -0.039060  0.166663 -0.060628   
3    0.0  0.0  0.0  0.072380  0.001230 -0.018210  0.133630 -0.037806   
4    0.0  0.0  0.0  0.087307 -0.013774 -0.015136  0.145934 -0.070759   
..   ...  ...  ...       ...       ...       ...       ...       ...   
205  0.0  0.0  0.0  0.089020 -0.004631 -0.040117  0.173419 -0.068980   
206  0.0  0.0  0.0  0.076448 -0.024823 -0.019830  0.134515 -0.093043   
207  0.0  0.0  0.0  0.051935  0.004076 -0.021560  0.104980 -0.037268   
208  0.0  0.0  0.0  0.075985 -0.012906 -0.024148  0.136739 -0.065944   
209  0.0  0.0  0.0  0.074415 -0.013738 -0.031449  0.138998 -0.070719   

          z_2       x_3  ...      x_18      y_18      z_18      x_19  \
0   -0.037758  0.157946  ...  0.015763 -0.236859 -0.053696  0.014897   
1   -0.049802  0.195749  ...  0.022846 -0.301951 -0.067628  0.019003   
2   -0.056427  0.204637  ...  0.011525 -0.315707 -0.082841  0.011519   
3   -0.024918  0.162663  ...  0.032818 -0.254179 -0.044583  0.030884   
4   -0.021705  0.181889  ...  0.044853 -0.331586 -0.057311  0.039559   
..        ...       ...  ...       ...       ...       ...       ...   
205 -0.058396  0.211823  ...  0.015739 -0.327877 -0.080195  0.015405   
206 -0.027526  0.156989  ...  0.006143 -0.281344 -0.046327  0.004280   
207 -0.027891  0.130062  ...  0.009148 -0.201698 -0.032603  0.009347   
208 -0.034568  0.169121  ...  0.014171 -0.263591 -0.055377  0.012828   
209 -0.044023  0.169419  ...  0.003324 -0.264689 -0.053460  0.004267   

         y_19      z_19      x_20      y_20      z_20  target  
0   -0.160986 -0.042879  0.003309 -0.143430 -0.028293       a  
1   -0.205834 -0.053242  0.004331 -0.182658 -0.034498       a  
2   -0.209255 -0.069801 -0.000003 -0.184926 -0.051607       a  
3   -0.179450 -0.034737  0.014321 -0.144736 -0.021350       a  
4   -0.234893 -0.044078  0.012328 -0.194194 -0.025516       a  
..        ...       ...       ...       ...       ...     ...  
205 -0.222312 -0.064478 -0.000278 -0.202003 -0.044425       a  
206 -0.202661 -0.035166 -0.000106 -0.154207 -0.019415       a  
207 -0.139441 -0.025672  0.004016 -0.114325 -0.014465       a  
208 -0.183512 -0.043389  0.003203 -0.158895 -0.027436       a  
209 -0.183576 -0.042859 -0.001866 -0.155362 -0.027300       a  

[210 rows x 64 columns]

In [6]:
df[df['target'] == 'b']

x_0  y_0  z_0       x_1       y_1       z_1       x_2       y_2  \
420  0.0  0.0  0.0  0.063047 -0.023764 -0.015633  0.111474 -0.126108   
421  0.0  0.0  0.0  0.077677 -0.029166 -0.024637  0.133963 -0.135797   
422  0.0  0.0  0.0  0.063846 -0.015011 -0.023965  0.114944 -0.101225   
423  0.0  0.0  0.0  0.062339 -0.023871 -0.021267  0.113165 -0.107567   
424  0.0  0.0  0.0  0.064134 -0.028756 -0.022028  0.113750 -0.114887   
..   ...  ...  ...       ...       ...       ...       ...       ...   
625  0.0  0.0  0.0  0.058419 -0.030207 -0.015475  0.099631 -0.108319   
626  0.0  0.0  0.0  0.060126 -0.017372 -0.013335  0.105420 -0.098992   
627  0.0  0.0  0.0  0.070404 -0.006071 -0.017204  0.135407 -0.109200   
628  0.0  0.0  0.0  0.068547 -0.012165 -0.014034  0.125481 -0.104795   
629  0.0  0.0  0.0  0.064126 -0.020897 -0.023634  0.111174 -0.106944   

          z_2       x_3  ...      x_18      y_18      z_18      x_19  \
420 -0.015273  0.117327  ...  0.018098 -0.390239 -0.032541  0.029245   
421 -0.028595  0.128716  ...  0.006046 -0.404690 -0.049162  0.017995   
422 -0.026450  0.109210  ...  0.022602 -0.338525 -0.024050  0.034710   
423 -0.023207  0.111182  ...  0.019820 -0.321470 -0.022635  0.030051   
424 -0.023899  0.109048  ...  0.016710 -0.330782 -0.018656  0.026637   
..        ...       ...  ...       ...       ...       ...       ...   
625 -0.016353  0.097771  ...  0.015726 -0.307548 -0.022291  0.024386   
626 -0.012475  0.108897  ...  0.029067 -0.332348 -0.025400  0.040817   
627 -0.017203  0.149792  ...  0.051218 -0.426427 -0.046215  0.069479   
628 -0.012689  0.135076  ...  0.048240 -0.398056 -0.035419  0.063773   
629 -0.024843  0.109917  ...  0.018873 -0.320796 -0.013422  0.030369   

         y_19      z_19      x_20      y_20      z_20  target  
420 -0.455837 -0.041731  0.036757 -0.515435 -0.048519       b  
421 -0.475040 -0.063089  0.028325 -0.541893 -0.073049       b  
422 -0.392095 -0.035076  0.044988 -0.442140 -0.043324       b  
423 -0.370892 -0.031228  0.038455 -0.418239 -0.038222       b  
424 -0.383943 -0.028698  0.034285 -0.432992 -0.036586       b  
..        ...       ...       ...       ...       ...     ...  
625 -0.356151 -0.029392  0.031174 -0.400676 -0.034954       b  
626 -0.387573 -0.033769  0.049316 -0.437535 -0.039949       b  
627 -0.499157 -0.057802  0.083838 -0.569521 -0.065745       b  
628 -0.464220 -0.046029  0.074907 -0.525973 -0.053582       b  
629 -0.371408 -0.024301  0.039568 -0.417912 -0.033064       b  

[210 rows x 64 columns]

In [7]:
df[df['target'] == 'c']

x_0  y_0  z_0       x_1       y_1       z_1       x_2       y_2  \
210  0.0  0.0  0.0  0.074740 -0.005406  0.011533  0.129689 -0.070941   
211  0.0  0.0  0.0  0.058287  0.002335  0.003196  0.103122 -0.042353   
212  0.0  0.0  0.0  0.070365 -0.011590  0.013311  0.122582 -0.086196   
213  0.0  0.0  0.0  0.067062 -0.013349  0.013964  0.116662 -0.081406   
214  0.0  0.0  0.0  0.064259 -0.009375  0.003520  0.116067 -0.072512   
..   ...  ...  ...       ...       ...       ...       ...       ...   
415  0.0  0.0  0.0  0.062432 -0.003917  0.001934  0.115769 -0.071857   
416  0.0  0.0  0.0  0.067867 -0.006935  0.015609  0.114691 -0.057831   
417  0.0  0.0  0.0  0.069791 -0.011537  0.012637  0.120119 -0.079101   
418  0.0  0.0  0.0  0.071314 -0.007569  0.014654  0.120373 -0.059978   
419  0.0  0.0  0.0  0.066761 -0.002139  0.002527  0.122181 -0.067481   

          z_2       x_3  ...      x_18      y_18      z_18      x_19  \
210  0.016830  0.164919  ...  0.097903 -0.379270 -0.033036  0.135918   
211  0.007224  0.129175  ...  0.072050 -0.293571 -0.014275  0.098791   
212  0.016351  0.161438  ...  0.090617 -0.394149 -0.055463  0.132981   
213  0.017737  0.151236  ...  0.087692 -0.372226 -0.050658  0.127769   
214  0.004304  0.148065  ...  0.062640 -0.366870 -0.032594  0.098788   
..        ...       ...  ...       ...       ...       ...       ...   
415  0.002907  0.148040  ...  0.060754 -0.374619 -0.035848  0.099967   
416  0.022275  0.145509  ...  0.083861 -0.335960 -0.028267  0.116998   
417  0.016036  0.155316  ...  0.086661 -0.374195 -0.053077  0.127038   
418  0.021179  0.152055  ...  0.092074 -0.346407 -0.030822  0.126158   
419  0.002612  0.157885  ...  0.067176 -0.374349 -0.042604  0.108648   

         y_19      z_19      x_20      y_20      z_20  target  
210 -0.362435 -0.037971  0.153537 -0.315532 -0.037863       c  
211 -0.282593 -0.019037  0.109129 -0.244201 -0.019453       c  
212 -0.363918 -0.059314  0.148968 -0.308942 -0.057448       c  
213 -0.342820 -0.055277  0.144033 -0.288455 -0.054999       c  
214 -0.353422 -0.037707  0.116162 -0.305608 -0.037380       c  
..        ...       ...       ...       ...       ...     ...  
415 -0.358185 -0.040228  0.116540 -0.308916 -0.038077       c  
416 -0.319782 -0.031475  0.133911 -0.276201 -0.030894       c  
417 -0.347239 -0.056947  0.142280 -0.294635 -0.055391       c  
418 -0.330693 -0.035048  0.140579 -0.284686 -0.035262       c  
419 -0.355112 -0.048795  0.128903 -0.303140 -0.047777       c  

[210 rows x 64 columns]

In [8]:
df.columns

Index(['x_0', 'y_0', 'z_0', 'x_1', 'y_1', 'z_1', 'x_2', 'y_2', 'z_2', 'x_3',
       'y_3', 'z_3', 'x_4', 'y_4', 'z_4', 'x_5', 'y_5', 'z_5', 'x_6', 'y_6',
       'z_6', 'x_7', 'y_7', 'z_7', 'x_8', 'y_8', 'z_8', 'x_9', 'y_9', 'z_9',
       'x_10', 'y_10', 'z_10', 'x_11', 'y_11', 'z_11', 'x_12', 'y_12', 'z_12',
       'x_13', 'y_13', 'z_13', 'x_14', 'y_14', 'z_14', 'x_15', 'y_15', 'z_15',
       'x_16', 'y_16', 'z_16', 'x_17', 'y_17', 'z_17', 'x_18', 'y_18', 'z_18',
       'x_19', 'y_19', 'z_19', 'x_20', 'y_20', 'z_20', 'target'],
      dtype='object')

In [10]:
list(df.columns[0:4])

['x_0', 'y_0', 'z_0', 'x_1']

# Autre option : load depuis csv en local

In [11]:
df = pd.read_csv('../csv/coord_live_abcdefghijklmnopqrstuvwxyz.csv', sep = ',' )
df.shape


(3900, 64)

In [12]:
df.head(3)

x_0  y_0  z_0       x_1       y_1       z_1       x_2       y_2       z_2  \
0  0.0  0.0  0.0  0.067701 -0.043059 -0.013783  0.116625 -0.112220 -0.018773   
1  0.0  0.0  0.0  0.061455 -0.026629 -0.019453  0.118439 -0.095989 -0.025251   
2  0.0  0.0  0.0  0.073134 -0.043769 -0.017741  0.129270 -0.115775 -0.023245   

        x_3  ...      x_18      y_18      z_18      x_19      y_19      z_19  \
0  0.139905  ...  0.002343 -0.251392 -0.030343  0.007545 -0.188142 -0.019205   
1  0.141798  ... -0.001150 -0.247533 -0.032880  0.004090 -0.179724 -0.023359   
2  0.154982  ...  0.003378 -0.277963 -0.024795  0.007469 -0.206901 -0.013835   

       x_20      y_20      z_20  target  
0  0.000437 -0.155193 -0.005219       a  
1 -0.001019 -0.142552 -0.009493       a  
2  0.001835 -0.172018  0.001015       a  

[3 rows x 64 columns]

## Features

In [13]:
X = df.drop(columns=['x_0', 'y_0', 'z_0', 'target'])
X.head(3)

x_1       y_1       z_1       x_2       y_2       z_2       x_3  \
0  0.067701 -0.043059 -0.013783  0.116625 -0.112220 -0.018773  0.139905   
1  0.061455 -0.026629 -0.019453  0.118439 -0.095989 -0.025251  0.141798   
2  0.073134 -0.043769 -0.017741  0.129270 -0.115775 -0.023245  0.154982   

        y_3       z_3       x_4  ...      z_17      x_18      y_18      z_18  \
0 -0.190496 -0.025808  0.150985  ... -0.012942  0.002343 -0.251392 -0.030343   
1 -0.176057 -0.030775  0.150593  ... -0.012048 -0.001150 -0.247533 -0.032880   
2 -0.206230 -0.030657  0.167805  ... -0.004927  0.003378 -0.277963 -0.024795   

       x_19      y_19      z_19      x_20      y_20      z_20  
0  0.007545 -0.188142 -0.019205  0.000437 -0.155193 -0.005219  
1  0.004090 -0.179724 -0.023359 -0.001019 -0.142552 -0.009493  
2  0.007469 -0.206901 -0.013835  0.001835 -0.172018  0.001015  

[3 rows x 60 columns]

## Formatage/encoding target y

In [14]:
y = df['target']
y.head(3)

0    a
1    a
2    a
Name: target, dtype: object

In [15]:
from sklearn.preprocessing import LabelEncoder

encoder = LabelEncoder()
y_encoded = encoder.fit_transform(y)
y_encoded

array([ 0,  0,  0, ..., 25, 25, 25])

## Proportion/distribution des classes

In [16]:
y.unique()

array(['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm',
       'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z'],
      dtype=object)

In [17]:
# Dispersion des classes dans le DF d'origine
round(df[['target']].groupby(df['target']).count().rename(columns={'target': '%'}) / len(df['target']), 2)

%
target      
a       0.04
b       0.04
c       0.04
d       0.04
e       0.04
f       0.04
g       0.04
h       0.04
i       0.04
j       0.04
k       0.04
l       0.04
m       0.04
n       0.04
o       0.04
p       0.04
q       0.04
r       0.04
s       0.04
t       0.04
u       0.04
v       0.04
w       0.04
x       0.04
y       0.04
z       0.04

## Modelisation

### Split

In [18]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y_encoded, test_size=.30, random_state=1, stratify=y)

### Check proportion/distibution dans Train/test

In [19]:
# Dispersion des classes dans y_train
y_train_temp = pd.DataFrame(y_train)
y_train_temp['%'] = 1
round(y_train_temp[['%']].groupby(y_train_temp[0]).count() / len(y_train_temp[['%']]), 2)

%
0       
0   0.04
1   0.04
2   0.04
3   0.04
4   0.04
5   0.04
6   0.04
7   0.04
8   0.04
9   0.04
10  0.04
11  0.04
12  0.04
13  0.04
14  0.04
15  0.04
16  0.04
17  0.04
18  0.04
19  0.04
20  0.04
21  0.04
22  0.04
23  0.04
24  0.04
25  0.04

In [20]:
# Dispersion des classes dans y_test
y_test_temp = pd.DataFrame(y_test)
y_test_temp['%'] = 1
round(y_test_temp[['%']].groupby(y_test_temp[0]).count() / len(y_test_temp[['%']]), 2)

%
0       
0   0.04
1   0.04
2   0.04
3   0.04
4   0.04
5   0.04
6   0.04
7   0.04
8   0.04
9   0.04
10  0.04
11  0.04
12  0.04
13  0.04
14  0.04
15  0.04
16  0.04
17  0.04
18  0.04
19  0.04
20  0.04
21  0.04
22  0.04
23  0.04
24  0.04
25  0.04

In [34]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, LSTM, Dropout
from tensorflow.keras.callbacks import EarlyStopping

# RESULTAT DE CE MODELE : ACCURACY @ 98% et LOSS @ 0.07, 25 minutes pour l'entrainement

# Remarque : Même modèle que pour les lettres A, B, C

# Avec la methode de collecte lag, on a de moins bon resultats : 96% pour loss @ 0.18

In [22]:
def search_best_model_params(X_train, y_train, X_test, y_test):
    thierry_model_LSTM_all = Sequential()

    thierry_model_LSTM_all.add(LSTM(128, return_sequences=True, activation='relu', input_shape=(60,1)))
    thierry_model_LSTM_all.add(LSTM(64, return_sequences=True, activation='relu'))
    thierry_model_LSTM_all.add(LSTM(32, return_sequences=True, activation='relu'))
    thierry_model_LSTM_all.add(LSTM(26, return_sequences=False, activation='relu'))
    #thierry_model_LSTM_all.add(Dropout(0.15))
    thierry_model_LSTM_all.add(Dense(26, activation='relu'))
    #thierry_model_LSTM_all.add(Dropout(0.3))
    thierry_model_LSTM_all.add(Dense(16, activation='relu'))
    thierry_model_LSTM_all.add(Dense(26, activation='softmax'))

    #model.compile(loss='mse', optimizer=Adam(learning_rate=0.5), metrics=['mae'])
    thierry_model_LSTM_all.compile(loss='sparse_categorical_crossentropy', optimizer='rmsprop', metrics=['accuracy'])

    es = EarlyStopping(patience=7, restore_best_weights=True)

    #model.compile(loss='mse', optimizer=Adam(learning_rate=0.5), metrics=['mae'])
    thierry_model_LSTM_all.compile(loss='sparse_categorical_crossentropy', optimizer='rmsprop', metrics=['accuracy'])

    es = EarlyStopping(patience=10, restore_best_weights=True)

    history = thierry_model_LSTM_all.fit(
        X_train, y_train, 
        #validation_data=(X_test, y_test),
        validation_split = 0.3,
        callbacks=[es],
        epochs=200, batch_size=32, verbose=0)

    params = thierry_model_LSTM_all.evaluate(X_test, y_test, verbose = 0)
    
    return params
    

[0.07886786758899689, 0.9794871807098389]

# Sauvegarde du modèle

In [23]:
import pickle

def upload_model(model, name_of_model):
    with open(f'{name_of_model}', 'wb') as file:
        pickle.dump(model, file)


def load_model(name_of_the_model=False):
    if name_of_the_model == False:
        with open('model.randomfo', 'rb') as file:
            loaded_model = pickle.load(file)
        return loaded_model
    else:
        with open(name_of_the_model, 'rb') as file:
            loaded_model = pickle.load(file)
        return loaded_model

In [24]:
#upload_model(thierry_model_LSTM, 'thierry_model_LSTM')
upload_model(thierry_model_LSTM_all, 'thierry_model_LSTM_live_all_98_no_verbose')

# Test avec d'autres dataset

In [25]:
#Il faut renseigner le nim du dataset ici

#df_verif = load_dataset(PROJECT_ID, 'test_dataset.test7')
df_verif = load_dataset(PROJECT_ID, 'test_dataset.test10')
#df_verif = pd.read_csv('../csv/coord_live_a_b.csv', sep = ',' )
#df_verif.head(3)

wagon-give-me-a-sign test_dataset.test10


In [26]:
X_verif = df_verif.drop(columns=['x_0', 'y_0', 'z_0', 'target'])
y_verif = df_verif['target']
y_verif_encoded = encoder.transform(y_verif)


In [27]:
#model_verif = load_model('thierry_model_LSTM_live')
model_verif2 = load_model('thierry_model_LSTM_live_all_98')
model_verif3 = load_model('thierry_model_LSTM_live_all_98_no_verbose')

In [33]:
model_verif2.evaluate(X_verif, y_verif_encoded, verbose = 0)

[0.21864818036556244, 0.9333333373069763]

In [32]:
model_verif3.evaluate(X_verif, y_verif_encoded, verbose = 0)

[0.6554323434829712, 0.7507936358451843]

# RESULTAT OK : ACCURACY @ 93% avec le dataset7 et le dataset10